sentence-embedding/python-lib/dku_language_model/context_independent_language_model.py

```diff
@@ -0,0 +1,176 @@
+
+    def get_sentence_word_vectors(self, text):
+        indices = [self.word2idx[w] for w in text.split() if w in self.word2idx]
+        return self.embedding_matrix[indices]
+
+    def compute_average_embedding(self, text):
+        embeddings = self.get_sentence_word_vectors(text)
+        avg_embedding = np.mean(embeddings, axis=0)
+        return avg_embedding
+        
+    def get_sentence_embedding(self, texts):
+        cleaned_texts = map(self.clean_text, texts)
+        res = map(self.compute_average_embedding, cleaned_texts)
+        return res
+        
+    def get_weighted_sentence_word_vectors(self, text, weights):
+        #Check if sentence contains at least one token and return None if not
+        indices = [self.word2idx[w] for w in text.split() if w in self.word2idx]
+        embeddings = self.embedding_matrix[indices]
+        weights = [weights[w] for w in text.split() if w in self.word2idx]
+        return [w * e for w, e in zip(weights, embeddings)]
+    
+    def compute_weighted_average_embedding(self, text, weights):
+        """Weighted average embedding for computing SIF."""
+        embeddings = self.get_weighted_sentence_word_vectors(text, weights)    
+        avg_embedding = np.mean(embeddings, axis=0)
+        return avg_embedding
+    
+    def remove_first_principal_component(self, X, npc):
+        """Removes the first PC for computing SIF."""
+        svd = TruncatedSVD(n_components=npc, n_iter=7, random_state=0)
+        X = np.array(X)
+        svd.fit(X)
+        u = svd.components_
+        return X - X.dot(u.T).dot(u)
+
+    def contruct_final_res(self, res,is_void):
+        res_final = []
+        j = 0
+        for v in is_void:
+            if v == 0:
+                res_final.append(res[j])
+                j+=1
+            else:
+                res_final.append(np.nan)
+        return res_final
+
+    def get_weighted_sentence_embedding(self, texts, smoothing_parameter, npc):
+        cleaned_texts = map(self.clean_text, texts)
+
+        # Compute word weights
+        word_weights = Counter()
+        for text in cleaned_texts:
+            word_weights.update(text.split())
+
+        n_words = float(sum(word_weights.values()))
+        for k, v in word_weights.items():
+            word_weights[k] /= n_words
+            word_weights[k] = float(smoothing_parameter) / (smoothing_parameter + word_weights[k])
+
+        # Compute SIF
+        logger.info("Computing weighted average embeddings...")
+        res = map(lambda s: self.compute_weighted_average_embedding(s, word_weights), cleaned_texts)
+
+        # Remove empty sentences and save their indecies
+        is_void = map(lambda x: 1 if x.shape == () else 0 , res)
+        res = [x for x,y in zip(res,is_void) if y==0]
+
+        logger.info("Removing vectors first principal component...")
+        res = self.remove_first_principal_component(res, npc)
+        res = self.contruct_final_res(res,is_void)
```

Comment:
This is really "tordu". I think line 83 changes the length of res, so when you enter the construct_final you need to increment j in a kind of smart way to not go out of bounds. Maybe don't do the list comp of line 83 and use an enumerate(is_void) in line 56 to keep track of the index.

Reply:
We can't really remove the list comprehension of line 83 because `remove_principal_component` needs to have a non-null array.
I rewrote the this part so that the logic is a bit cleaner

Reply:
Update: if I want to make it clearer, I need to touch some of the upper part (to make the `res` array more numpy friendly), not sure we want to to it right now.